test/integration/emulator/TimeTurner.cpp

```diff
@@ -0,0 +1,143 @@
+#include "TimeTurner.h"
+#include "logger/Logger.h"
+#include <thread>
+
+namespace emulator
+{
+
+TimeTurner::TimeTurner() : _timestamp(100), _startTime(std::chrono::system_clock::now()), _running(true) {}
+
+void TimeTurner::nanoSleep(const uint64_t nanoSeconds)
+{
+    if (!_running)
+    {
+        std::this_thread::yield();
+        return;
+    }
+
+    const auto timerExpires = _timestamp + nanoSeconds;
+    for (size_t i = 0; i < _sleepers.size() * 5; ++i)
+    {
+        auto& slot = _sleepers[i % _sleepers.size()];
+        auto expectedState = State::Empty;
+        if (slot.state.compare_exchange_strong(expectedState, State::Allocated))
+        {
+            slot.expireTimestamp = timerExpires;
+            slot.state.store(State::Committed);
+            _sleeperCount.post();
+            slot.semaphore.wait();
+
+            slot.state.store(State::Empty);
+            return;
+        }
+    }
+
+    assert(false);
+}
+
+std::chrono::system_clock::time_point TimeTurner::wallClock()
+{
+    return _startTime + std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::nanoseconds(_timestamp));
+}
+
+/**
+ * You can run this initially if you know how many threads minimum will join the sleep.
+ * It can only be run once.
+ */
+void TimeTurner::waitForThreadsToSleep(uint32_t expectedCount, uint64_t timeoutNs)
+{
+    for (uint32_t i = 0; i < expectedCount - 1; ++i)
+    {
+        _sleeperCount.decrement();
```

Comment:
Now I guess I understand the changes on semaphore, although I think this logic not be archived by a weird implementation of a semaphore. My suggestion is implement that logic here or other type that manage int value and a method that can wait for a specific condition on that value

Reply:
I will create another one. It is just that a sempahore is logically a resource counter. When there are resources available, consumers can take one and continue until the resources are depleted and you will have to wait for resources to become available again.
The semaphore I am after is the same except that it starts with a debt of resources, instead of 0 resources.